python/uv/private/uv.bzl

```diff
@@ -18,36 +18,366 @@ EXPERIMENTAL: This is experimental and may be removed without notice
+            config.setdefault("urls", {})[platform] = struct(
+                sha256 = sha256,
+                urls = urls,
+            )
+    elif compatible_with or target_settings:
+        fail("`platform` name must be specified when specifying `compatible_with` or `target_settings`")
+
+def parse_modules(module_ctx, uv_repository = None):
+    """Parse the modules to get the config for 'uv' toolchains.
+
+    Args:
+        module_ctx: the context.
+        uv_repository: the rule to create a uv_repository override.
+
+    Returns:
+        A dictionary for each version of the `uv` to configure.
+    """
+    defaults = {}
     for mod in module_ctx.modules:
-        for toolchain in mod.tags.toolchain:
-            if not mod.is_root:
-                fail(
-                    "Only the root module may configure the uv toolchain.",
-                    "This prevents conflicting registrations with any other modules.",
-                    "NOTE: We may wish to enforce a policy where toolchain configuration is only allowed in the root module, or in rules_python. See https://github.com/bazelbuild/bazel/discussions/22024",
-                )
-
-            uv_repositories(
-                uv_version = toolchain.uv_version,
-                register_toolchains = False,
+        for default_attr in mod.tags.default:
+            _configure(
+                defaults,
+                version = default_attr.version,
+                base_url = default_attr.base_url,
+                manifest_filename = default_attr.manifest_filename,
+                platform = default_attr.platform,
+                compatible_with = default_attr.compatible_with,
+                target_settings = default_attr.target_settings,
+            )
+
+    versions = {}
+    for mod in module_ctx.modules:
+        last_version = None
+        for config_attr in mod.tags.configure:
+            last_version = config_attr.version or last_version or defaults["version"]
+            if not last_version:
+                fail("version must be specified")
+
+            specific_config = versions.setdefault(
+                last_version,
+                {
+                    "base_url": defaults.get("base_url", ""),
+                    "manifest_filename": defaults["manifest_filename"],
+                    "platforms": dict(defaults["platforms"]),  # copy
+                },
+            )
+
+            _configure(
+                specific_config,
+                base_url = config_attr.base_url,
+                manifest_filename = config_attr.manifest_filename,
+                platform = config_attr.platform,
+                compatible_with = config_attr.compatible_with,
+                target_settings = config_attr.target_settings,
+                sha256 = config_attr.sha256,
+                urls = config_attr.urls,
             )
 
+    versions = {
+        version: config
+        for version, config in versions.items()
+        if config["platforms"]
```

Comment:
I don't get why this guard is here -- why does an empty platforms occur, and why does it need to be filtered out?

Reply:
Will remove. Not needed.